cpp/src/session.cpp

```diff
@@ -148,4 +158,209 @@ void* session::user_data() const {
+
+void transaction_handle_outcome(const session& s, proton::tracker t) {
+    session_context::get(unwrap(s))._txn_impl->handle_outcome(t);
+}
+
+void transaction_delete(const session& s) { auto &_txn_impl = session_context::get(unwrap(s))._txn_impl; delete _txn_impl; _txn_impl = nullptr;}
+
+}
+
+void session::transaction_declare(proton::messaging_handler &handler, bool settle_before_discharge) {
+    auto &txn_impl = session_context::get(pn_object())._txn_impl;
+    if (txn_impl == nullptr) {
+        // Create _txn_impl
+        proton::connection conn = this->connection();
+        class InternalTransactionHandler : public proton::messaging_handler {
+
+            void on_tracker_settle(proton::tracker &t) override {
+                if (!transaction_is_empty(t.session())) {
+                    transaction_handle_outcome(t.session(), t);
+                }
+            }
+        };
+
+        proton::target_options opts;
+        std::vector<symbol> cap = {proton::symbol("amqp:local-transactions")};
+        opts.capabilities(cap);
+        opts.make_coordinator();
+
+        proton::sender_options so;
+        so.name("txn-ctrl");
+        so.target(opts);
+
+        static InternalTransactionHandler internal_handler; // internal_handler going out of scope. Fix it
+        so.handler(internal_handler);
+
+        static proton::sender s = conn.open_sender("does not matter", so);
+
+        settle_before_discharge = false;
+
+        txn_impl = new transaction_impl(s, handler, settle_before_discharge);
+    }
+    // Declare txn
+    txn_impl->declare();
+}
+
+
+proton::binary session::transaction_id() const { return session_context::get(pn_object())._txn_impl->transaction_id; }
+void session::transaction_commit() { session_context::get(pn_object())._txn_impl->commit(); }
+void session::transaction_abort() { session_context::get(pn_object())._txn_impl->abort(); }
+bool session::transaction_is_declared() { return (!transaction_is_empty(*this)) && session_context::get(pn_object())._txn_impl->state == transaction_impl::State::DECLARED; }
+
+transaction_impl::transaction_impl(proton::sender &_txn_ctrl,
+                                   proton::messaging_handler &_handler,
+                                   bool _settle_before_discharge)
+    : txn_ctrl(_txn_ctrl), handler(&_handler) {
+}
+transaction_impl::~transaction_impl() {}
+
+void transaction_impl::commit() {
+    discharge(false);
+}
+
+void transaction_impl::abort() {
+    discharge(true);
+}
+
+void transaction_impl::declare() {
+    if (state != transaction_impl::State::FREE)
+        throw proton::error("This session has some associcated transaction already");
+    state = State::DECLARING;
+
+    proton::symbol descriptor("amqp:declare:list");
+    std::list<proton::value> vd;
+    proton::value i_am_null;
+    vd.push_back(i_am_null);
+    proton::value _value = vd;
+    send_ctrl(descriptor, _value);
```

Comment:
Can't this be something shorter using inline construction? In any case get rid of `_` leading `_value`